### ISSUE TITLE
Add support for RFC 0557: Discover Features v2

### ISSUE THREAD

**Issue by TimoGlastra** [hyperledger/aries-framework-javascript#495]
https://github.com/hyperledger/aries-rfcs/tree/b3a3942ef052039e73cd23d847f42947f8287da2/features/0557-discover-features-v2

**Comment by genaris** [comment]
It looks like this is will have some priority in the short term, as it is required for AIP 2.0 compatibility and also it becomes more important as framework is growing (as shown in #956). Is there any ongoing work on this issue?

It's not so trivial to implement, as currently Discover Features module is using Agent Dispatcher to tell which protocols it supports and this is limited to message types. But probably a sort of _ProtocolRegistry_ can be created in order to let each protocol module to register its capabilities (in terms of protocols, roles and goal codes), and then query it by the Discover Features module, and maybe this wouldn't introduce breaking changes and even be incorporated in 0.2.x releases.

**Comment by TimoGlastra** [comment]
I don't think anyone is working on this, so feel free to pick this up! :)

Yeah I think we should have some sort of protocol registry. We currently register the handlers on the dispatcher, but if we extract that to a separate protocol registry where you can register handlers (for message types) but also decorators, roles for a protocol and other items so it can be extended beyond the current known discovery types. 

I was first thinking of extending the handlers, but a handler is just one message, so not sure we should add items to it that are on the protocol lever instead of the message level

**Comment by genaris** [comment]
I kept thinking about this these days and my current idea is to define a 'FeatureRegistry' with the capabilities of registering items (features) you mentioned, that will be queried by DiscoveryFeatures module when some agent.

A good entry point for that feature registration can be the Module Registration API that will be fully available for 0.3.0, so I was thinking on going straight there. However, we were discussing briefly about this in today's call with @JamesKEbert and, as there is interest on supporting this soon, we can try to find a way of introducing in 0.2.x first.

So what I'm wondering now is that each module can register its features in `register`method, which would take not only the DependencyManager as parameter but also this new FeatureRegistry. Something like: 

```ts
public static register(featureRegistry: FeatureRegistry, dependencyManager: DependencyManager) {
    // Api
    dependencyManager.registerContextScoped(Module)

    // Services
    dependencyManager.registerSingleton(Service)

    // Features
    featureRegistry.registerFeatures([protocols, roles, goal-codes,,etc.])
  }
}
```

This way I guess it will not break anything (as this registration API is internal/experimental) and would give `register` method a wider scope than just registering dependencies.

However, DiscoverFeatures module will depend on the FeatureRegistry, so I think its constructor would need to be changed. Would that be considered a breaking change? 😱 If that's the case, it could be managed as a separate module (i.e. `DiscoverFeaturesV2Module`) for the time being.

What do you guys think about these alternatives?

**Comment by TimoGlastra** [comment]
> However, DiscoverFeatures module will depend on the FeatureRegistry, so I think its constructor would need to be changed. Would that be considered a breaking change? 😱 If that's the case, it could be managed as a separate module (i.e. DiscoverFeaturesV2Module) for the time being.

Why would it need to be changed?

---

If this is a desired feature for 0.2.x then I think the aproach as you mentioned seems fine. Even for 0.3.0 it seems like a fine approach to me. 👍
